src/slack-request-signature-validator.ts

```diff
@@ -0,0 +1,53 @@
+import crypto from "crypto";
+import { Request } from "express";
+import { inject, injectable } from "inversify";
+import timingSafeCompare from "tsscmp";
+
+import { SecretProvider } from "./secret-provider";
+
+@injectable()
+class SlackRequestSignatureValidator {
```

Comment:
The method `isSignatureValid()` is doing a lot of things so I think it needs to be split up.

The requestSignature and requestTimestamp being set seem to be core requirements so perhaps these can be set & checked in the constructor?

Ideally I would want to see something like this:

```typescript
async function isSignatureValid(): Promise<boolean> {
    if (!this.isTimeStampValid()) { 
        return false;
    }
    const slackSigningSecret = await this.getSlackSigningSecret();
    const hmac = this.createSlackHmac(slackSigningSecret);
    const localSignature = this.createLocalSignature(hmac);

    return timingSafeCompare(this.requestSignature, localSignature);
}
```

Reply:
You raise quite a good point. An instance of a `SlackRequestSignatureValidator` does not need to cache an the entire `Request`. Have refactored to:

```typescript
  public async isSignatureValid(): Promise<boolean> {
    if (!this.requestSignature || !this.requestTimestamp) {
      return false;
    }

    if (!this.isRecentRequest()) {
      return false;
    }

    if (
      !timingSafeCompare(this.requestSignature, await this.hashRawRequestBody())
    ) {
      return false;
    }

    return true;
  }
```